server/src/com/cloud/storage/snapshot/SnapshotManagerImpl.java

```diff
@@ -1123,6 +1150,39 @@ public SnapshotInfo takeSnapshot(VolumeInfo volume) throws ResourceAllocationExc
         return snapshot;
     }
 
+    protected class BackupSnapshotTask extends ManagedContextRunnable {
+        SnapshotInfo snapshot;
+        int maxAttempts;
+        SnapshotStrategy snapshotStrategy;
+
+        public BackupSnapshotTask(SnapshotInfo snap, int maxRetries, SnapshotStrategy strategy) {
+            snapshot = snap;
+            maxAttempts = maxRetries;
+            snapshotStrategy = strategy;
+        }
+
+        @Override
+        protected void runInContext() {
+            try {
+                s_logger.debug("Value of attempts is " + maxAttempts);
+
+                SnapshotInfo backupedSnapshot = snapshotStrategy.backupSnapshot(snapshot);
+
+                if (backupedSnapshot != null) {
+                    snapshotStrategy.cleanSnapshotAfterCreation(snapshot);
+                }
+            } catch (final Exception e) {
+                if (maxAttempts >= 0) {
+                    s_logger.debug("Attempt failed, in retrying backup");
```

Comment:
Add more information in the log message, like backup attempt failed for which snapshot, how many attempts left for retry etc.